oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,459 @@
+    Map<String, List<String>> requestMetadata = super.getRequestMetadata(uri);
+    return addQuotaProjectIdToRequestMetadata(quotaProjectId, requestMetadata);
+  }
+
+  /**
+   * Returns credentials defined by a JSON file stream.
+   *
+   * <p>Returns {@link IdentityPoolCredentials} or {@link AwsCredentials}.
+   *
+   * @param credentialsStream the stream with the credential definition
+   * @return the credential defined by the credentialsStream
+   * @throws IOException if the credential cannot be created from the stream
+   */
+  public static ExternalAccountCredentials fromStream(InputStream credentialsStream)
+      throws IOException {
+    return fromStream(credentialsStream, OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+  }
+
+  /**
+   * Returns credentials defined by a JSON file stream.
+   *
+   * <p>Returns a {@link IdentityPoolCredentials} or {@link AwsCredentials}.
+   *
+   * @param credentialsStream the stream with the credential definition
+   * @param transportFactory the HTTP transport factory used to create the transport to get access
+   *     tokens
+   * @return the credential defined by the credentialsStream
+   * @throws IOException if the credential cannot be created from the stream
+   */
+  public static ExternalAccountCredentials fromStream(
+      InputStream credentialsStream, HttpTransportFactory transportFactory) throws IOException {
+    checkNotNull(credentialsStream);
+    checkNotNull(transportFactory);
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    GenericJson fileContents =
+        parser.parseAndClose(credentialsStream, StandardCharsets.UTF_8, GenericJson.class);
+    try {
+      return fromJson(fileContents, transportFactory);
+    } catch (ClassCastException e) {
+      throw new CredentialFormatException("An invalid input stream was provided.", e);
+    }
+  }
+
+  /**
+   * Returns external account credentials defined by JSON using the format generated by gCloud.
+   *
+   * @param json a map from the JSON representing the credentials
+   * @param transportFactory HTTP transport factory, creates the transport used to get access tokens
+   * @return the credentials defined by the JSON
+   */
+  static ExternalAccountCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) {
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
+    String subjectTokenType = (String) json.get("subject_token_type");
+    String tokenUrl = (String) json.get("token_url");
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+
+    Map<String, Object> credentialSourceMap = (Map<String, Object>) json.get("credential_source");
+
+    // Optional params.
+    String tokenInfoUrl =
+        json.containsKey("token_info_url") ? (String) json.get("token_info_url") : null;
```

Comment:
shouldn't json.get simply return null if there's no such key? Are these containsKey calls necessary?

If they are necessary (because this map doesn't allow null keys) then lins 229-232 need these checks too.

Reply:
removed the containsKey